terraform/main.tf

```diff
@@ -0,0 +1,199 @@
+terraform {
+    required_version = ">= 0.13"
+
+    required_providers {
+        azurerm = {
+            source  = "hashicorp/azurerm"
+            version = "2.25.0"
+        }
+        external = {
+            source  = "hashicorp/external"
+            version = "1.2.0"
+        }
+    }
+}
+
+provider "azurerm" {
+    version = "~> 2.0"
+    features {}
+}
+
+provider "external" {
+  version = "~> 1.2"
+}
+
+# Get info about currently activated subscription
+data "azurerm_subscription" "current" {}
+
+# Resource Group
+resource "azurerm_resource_group" "rg" {
+    name     = "Hub23"
+    location = "westeurope"
+}
+
+# Container Registry
+resource "azurerm_container_registry" "acr" {
+    name                = "hub23registry"
+    resource_group_name = azurerm_resource_group.rg.name
+    location            = azurerm_resource_group.rg.location
+    sku                 = "Standard"
+}
+
+# DNS Zone
+resource "azurerm_dns_zone" "dns" {
+    name                = "hub23.turing.ac.uk"
+    resource_group_name = azurerm_resource_group.rg.name
+}
+
+# A Records in DNS Zone
+resource "azurerm_dns_a_record" "binder_a_rec" {
+    name                = "binder"
+    zone_name           = azurerm_dns_zone.dns.name
+    resource_group_name = azurerm_resource_group.rg.name
+    ttl                 = 300
+    target_resource_id  = azurerm_public_ip.ipaddr.id
+}
+
+resource "azurerm_dns_a_record" "hub_a_rec" {
+    name                = "hub"
+    zone_name           = azurerm_dns_zone.dns.name
+    resource_group_name = azurerm_resource_group.rg.name
+    ttl                 = 300
+    target_resource_id  = azurerm_public_ip.ipaddr.id
+}
+
+# IP Address
+resource "azurerm_public_ip" "ipaddr" {
+    name                = "kubernetes-a5bd8ff872541442ca741cad811020fb"
+    resource_group_name = azurerm_kubernetes_cluster.k8s.node_resource_group
+    location            = azurerm_resource_group.rg.location
+    allocation_method   = "Static"
+    sku                 = "Standard"
+
+    tags = {
+        kubernetes-cluster-name = "kubernetes"
+        service                 = "hub23/hub23-nginx-ingress-controller"
+    }
+}
+
+# Key Vault
+resource "azurerm_key_vault" "keyvault" {
+    name                       = "hub23-keyvault"
+    location                   = azurerm_resource_group.rg.location
+    resource_group_name        = azurerm_resource_group.rg.name
+    sku_name                   = "standard"
+    tenant_id                  = data.azurerm_subscription.current.tenant_id
+    soft_delete_enabled        = true
+    soft_delete_retention_days = 90
+}
+
+# Kubernetes Cluster
+resource "azurerm_kubernetes_cluster" "k8s" {
+    name                = "hub23cluster"
+    kubernetes_version  = "1.16.15"
```

Comment:
I think version 1.16.x will stop being supported on AKS in Jan 2021. Might be worth upgrading here?
I found this stopped me being able to add new node pools etc... and eventually Azure might upgrade it for you (https://docs.microsoft.com/en-us/azure/aks/supported-kubernetes-versions)